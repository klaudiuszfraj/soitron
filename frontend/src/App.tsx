import React, {useEffect, useState} from 'react';
import Company from "./components/company/Company";
import {requestAPI} from "./utils/utils";
import Header from "./components/header/Header";
import {CompanyDBType} from "../../common/types";
import style from './App.module.scss'
import Search from "./components/search/Search";
import Dropdown from "./components/dropdown/Dropdown";

function App() {
    const [companies, setCompanies] = useState<(CompanyDBType & { id: number })[]>([])
    const [displayedCompanies, setDisplayedCompanies] = useState<(CompanyDBType & { id: number })[]>([])
    const [searchPhrase, setSearchPhrase] = useState('')
    const [filterServices, setFilterServices] = useState<string[]>([])

    useEffect(() => {
        requestAPI('/companies').then(companiesArray => {
            setCompanies(companiesArray)
            setDisplayedCompanies(companiesArray)
        })
    }, [])
    useEffect(() => {
        const regEx = new RegExp(searchPhrase, 'i')
        const searchedCompanies = companies.filter(company => company.name.match(regEx))
        const filteredCompanies = searchedCompanies.filter(company => filterServices.every(elem => company.services.includes(elem)))
        setDisplayedCompanies(filteredCompanies)
    }, [searchPhrase, filterServices, companies])
    return (
        <div className="App">
            <Header/>
            <main className={style.main}>
                {!companies.length && "Loading..."}
                <Search setSearchPhrase={setSearchPhrase}/>
                <Dropdown setFilterServices={setFilterServices} filterServices={filterServices}/>

                <section className={style.list}>
                    {displayedCompanies.length !== 0 && displayedCompanies.map(({id, name, logo, services, country}) =>
                        <Company key={id} name={name} logo={logo} services={services} country={country}/>
                    )}
                </section>

            </main>
        </div>
    );
}

export default App;
