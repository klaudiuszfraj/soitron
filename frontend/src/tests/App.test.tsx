import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import App from '../App';

test('renders loading page', async () => {
  render(<App />);
  const linkElement = await screen.findByText(/Agora/i);
  expect(linkElement).toBeInTheDocument();
});
test('search box', async () => {
  render(<App />);
  const searchBox = await screen.findByRole('textbox');
  expect(searchBox).toBeInTheDocument();
  await userEvent.type(searchBox, 'agor');
  const noResults = await screen.findByText(/agora/i);
  expect(noResults).toBeInTheDocument();
});
test('dropdown', async () => {
  render(<App />);
  const optionList = await screen.findAllByRole('list');
  expect(optionList).toMatchSnapshot(`[<ul class="options "><li class="option  ">Payroll</li><li class="option  ">Treasury</li><li class="option  ">Implementation</li><li class="option  ">Bank Payments</li></ul>]`);
  const selectServices = await screen.findByText('Bank Payments');
  await userEvent.click(selectServices);
  const filteredServices = await screen.findAllByText('Bank Payments');
  expect(filteredServices).toMatchSnapshot(`[<button class="optionBadge">Bank Payments<span class="removeBtn">Ã—</span></button>, <li class="option selected highlighted">Bank Payments</li>]`);


});
