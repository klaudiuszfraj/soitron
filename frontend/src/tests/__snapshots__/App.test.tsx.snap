// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dropdown: [<button class="optionBadge">Bank Payments<span class="removeBtn">×</span></button>, <li class="option selected highlighted">Bank Payments</li>] 1`] = `
Array [
  <button
    class="optionBadge"
  >
    Bank Payments
    <span
      class="removeBtn"
    >
      ×
    </span>
  </button>,
  <li
    class="option selected highlighted"
  >
    Bank Payments
  </li>,
]
`;

exports[`dropdown: [<ul class="options "><li class="option  ">Payroll</li><li class="option  ">Treasury</li><li class="option  ">Implementation</li><li class="option  ">Bank Payments</li></ul>] 1`] = `
Array [
  <ul
    class="options "
  >
    <li
      class="option  "
    >
      Payroll
    </li>
    <li
      class="option  "
    >
      Treasury
    </li>
    <li
      class="option  "
    >
      Implementation
    </li>
    <li
      class="option  "
    >
      Bank Payments
    </li>
  </ul>,
]
`;
